About/Heads-threat-model.md

```diff
@@ -0,0 +1,499 @@
+ ran "too late" -- the system has already fetched the kernel and initrd from the
+ disk and potentially had the chain of trust compromised. Since my version of
+ code is executed from the difficult-to-write SPI flash ROM and the read-only
+ boot block initializes the root of trust with measurements of itself as well as
+ the rest of the ROM, it is much harder to compromise.
+
+Additionally, since my ROM image is very size constrained, Heads didn't want to use
+ OpenSSL and liboath and all of the other dependencies. My branch replaces them
+ with mbedtls and my own TOTP code, which reduces the size of the executables
+ from 5MB to 180KB.  The source is available from [github.com/osresearch/tpmtotp](https://github.com/osresearch/tpmtotp)
+ and has since been merged into the Heads project.
+
+The UEFI firmware itself is also of great concern: is a very large code base and
+ most system firmwares are built from closed-source forks of the edk2 tree. This
+ presents a problem for trusted systems since it is not possible to knows what
+ else was included in the ROM image. Some vendors ship malware in the firmware
+ like Computrace
+ [Good Software can go Bad, Kaspersky 2014](https://www.kaspersky.com/about/press-releases/2014_good-software-can-go-bad)
+ and others bundle rootkits into the boot rom
+ <!-- markdownlint-disable MD013 -->
+ [Lenovo caught using rootkit to secretly install unremovable software, Khandelwal 2015](https://thehackernews.com/2015/08/lenovo-rootkit-malware.html)
+ <!-- markdownlint-enable MD013 -->
+ . Some vendors provide signed hashes for validating that the firmware
+ hasn't been tampered with, but this is just as opaque to the end user
+ <!-- markdownlint-disable MD013 -->
+ [Baked-in Lenovo Trusted Platform Assurance Helps Establish a Secure Foundation for Workloads (Jill Caugherty, 2015)](https://web.archive.org/web/20180415020314/http://blog.lenovo.com/en/blog/baked-in-lenovo-trusted-platform-assurance-establishes-a-secure-foundation/)
+  <!-- markdownlint-enable MD013 -->
+ .
+
+Additionally there are many other things that we would like to do before handing
+ control over to the OS on the harddrive and would like to develop them in a
+ comfortable open source environment. CoreBoot addresses both of these concerns
+ this by building a Linux kernel that is flashed into the system's boot ROM in
+ place of the UEFI image. There is still a need for some closed-source blobs
+ (Intel's MRC and ME firmware, for instance), but the bulk of the vendor
+ provided code is gone. Instead of a boot loader like Grub, Coreboot can invoke
+ a Linux kernel stored in ROM with its own immutable ROM file system. This
+ kernel and runtime is the bulk of Heads and is responsible for implementing the
+ rest of the verification steps like tpmtotp and verifying the next stages
+ before kexec()'ing the real OS.
+
+Finally, once Coreboot has been flashed into the ROM, the write protect pins on
```

Comment:
@tlaurion I am a little confused by this comment.  The paragraph talks about physically grounding the write pin to prevent write locks.  I can mention flashrom's write protect, but feel like specifying wp-ranges for specific chips in this way would be confusing and potentially dangerous.

Reply:
@Thrilleratplay absolutely. I was just quoting the doc at https://web.archive.org/web/20160927161825/http://www.tcij.org:80/resources/guides-0/replacing-BIOS so that there is a link somewhere to what Trammel refers to locking down WP bits, if it matches user's threat model. It is cross referenced in docs, but not referred directly anywhere. That would be the software version of it, expecting the SPI chip enforces the protection correctly, while grounding the pin to prevent write is the more radical option for some, where even external reprogramming would be prevented unless the ground is undid.

Reply:
@tlaurion For now, I want to leave it as is.  I am not sure what would happen if write protecting a `MX25L6405,MX25L6405D, MX25L6406E/MX25L6408E, MX25L6436E/MX25L6445E/MX25L6465E/MX25L6473E/MX25L6473F, MX25L6495F` and specify the wrong one.  Or if resizing the ME partition, if write protecting an area that is redefined in the IFD layout.  I am sure that flashrom has protections to prevent the worst from happening but would prefer to test this out myself and having a much more robust section explaining write protection, undoing write protection, steps to take if a mistake is made, etc..

Reply:
@Thrilleratplay Sorry. What I meant is to point to  https://web.archive.org/web/20160927161825/http://www.tcij.org:80/resources/guides-0/replacing-BIOS for all parts where WP bits are talked about (so devels/users know what we are talking about to experiment with)

Reply:
@Thrilleratplay In that case, please add a Write protection section, with WP bits and WP grounding subsections empty, linking at least to  https://web.archive.org/web/20160927161825/http://www.tcij.org:80/resources/guides-0/replacing-BIOS so that it is not lost (and others may want to contribute and test).